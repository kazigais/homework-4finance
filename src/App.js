import React, { PureComponent } from 'react';
import { connect } from 'react-redux';
import {
  Header, Button, Segment, Grid, Divider, Modal,
} from 'semantic-ui-react';
import moment from 'moment';
import Fade from 'react-reveal/Fade';
import PropTypes from 'prop-types';
import styles from './App.module.scss';
import {
  getLoanDetails, requestLoan, extendLoan, getLoanHistory,
} from './actions/loan';
import { toggleModal } from './actions/app';
import Calculator from './components/Calculator/Calculator';
import LoanHistory from './components/LoanHistory/LoanHistory';

export class App extends PureComponent {
  static propTypes = {
    app: PropTypes.shape({
      state: PropTypes.bool,
      message: PropTypes.string,
      modalState: PropTypes.bool,
      modalTitle: PropTypes.string,
      modalBody: PropTypes.string,
    }),
    loan: PropTypes.shape({
      repaymentAmount: PropTypes.number,
      amount: PropTypes.number,
      date: PropTypes.instanceOf(Date),
      apr: PropTypes.number,
    }),
    loanHistory: PropTypes.arrayOf(PropTypes.shape({

    })),
    getLoanDetails: PropTypes.func,
    requestLoan: PropTypes.func,
    extendLoan: PropTypes.func,
    toggleModal: PropTypes.func,
    getLoanHistory: PropTypes.func,
  }

  static defaultProps = {
    app: {},
    loan: {},
    loanHistory: [],
    getLoanDetails: () => {},
    requestLoan: () => {},
    extendLoan: () => {},
    toggleModal: () => {},
    getLoanHistory: () => {},
  }

  componentDidMount() {
    this.props.getLoanHistory();
  }

  render() {
    const { loan, app, loanHistory } = this.props;

    return (
      <div className={styles.container}>
        <Header size="huge">Micro-loan application</Header>
        <Segment className={styles['app-segment']} placeholder>
          <Grid columns={2} stackable textAlign="center">
            <Divider vertical />

            <Grid.Row verticalAlign="middle">
              <Grid.Column className={styles['segment-column']}>
                <Calculator
                  date={new Date(Date.now())}
                  title="Fixed term loan"
                  onSubmit={(amount, date) => this.props.getLoanDetails(amount, date, 10)}
                  errorMessage={app.state ? app.message : undefined}
                />
              </Grid.Column>

              <Grid.Column className={`${styles['segment-column']} ${styles.summary}`}>
                <Fade delay={500} collapse top when={loan.repaymentAmount !== undefined}>
                  <>
                    <Header className={styles['summary-title']} as="h2">
Total amount repayable: $
                      {loan.repaymentAmount}
                    </Header>
                    <Header className={styles['summary-title']} as="h2">
Loan cost: $
                      {(loan.repaymentAmount - loan.amount).toFixed(2)}
                    </Header>
                    <Header className={styles['summary-title']} as="h2">
Repayment date:
                      {loan.repaymentAmount !== undefined ? moment(loan.date).format('LL') : null}
                    </Header>
                    <Button
                      onClick={
                        () => this.props.requestLoan(loan.amount, loan.repaymentAmount, loan.date, loan.apr)
                      }
                      primary
                    >
                      Request loan
                    </Button>
                  </>
                </Fade>
                <Fade collapse top when={loan.repaymentAmount === undefined}>
                  <Header className={styles['summary-title']} as="h2">Fill out your required loan details in the calculator</Header>
                </Fade>
              </Grid.Column>
            </Grid.Row>
          </Grid>
        </Segment>
        <LoanHistory
          loanHistory={loanHistory}
          onLoanExtend={(id) => this.props.extendLoan(id)}
        />
        <Modal
          open={app.modalState}
          header={app.modalTitle}
          content={app.modalBody}
          onClose={() => this.props.toggleModal(false, null, null)}
        />
      </div>
    );
  }
}

const mapStateToProps = (state) => ({
  loan: state.loan,
  app: state.app,
  loanHistory: state.loanHistory,

});

const mapDispatchToProps = {
  getLoanDetails,
  requestLoan,
  extendLoan,
  toggleModal,
  getLoanHistory,
};

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(App);
