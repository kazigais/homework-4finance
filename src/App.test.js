import React from 'react';
import ReactDOM from 'react-dom';
import { App } from './App';
import { shallowToJson } from 'enzyme-to-json';
import { shallow } from 'enzyme';

describe('App component', () => {

  const props = {
    app: {},
    loanHistory: [],
    loan: {},
    getLoanHistory: () => {},
  }

  let dateNowSpy;

  beforeAll(() => {
      // Lock Time
      dateNowSpy = jest.spyOn(Date, 'now').mockImplementation(() => 1487076708000);
  });

  afterAll(() => {
      // Unlock Time
      dateNowSpy.mockRestore();
  });

  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<App {...props} />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('matches snapshot', () => {
    const wrapper = shallow(<App {...props} />);
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });
});
