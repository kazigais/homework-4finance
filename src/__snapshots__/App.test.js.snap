// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component matches snapshot 1`] = `
<div
  className="container"
>
  <Header
    size="huge"
  >
    Micro-loan application
  </Header>
  <Segment
    className="app-segment"
    placeholder={true}
  >
    <Grid
      columns={2}
      stackable={true}
      textAlign="center"
    >
      <Divider
        vertical={true}
      />
      <GridRow
        verticalAlign="middle"
      >
        <GridColumn
          className="segment-column"
        >
          <Calculator
            date={2017-02-14T12:51:48.000Z}
            onSubmit={[Function]}
            title="Fixed term loan"
          />
        </GridColumn>
        <GridColumn
          className="segment-column summary"
        >
          <Fade
            collapse={true}
            delay={500}
            top={true}
            when={false}
          >
            <Header
              as="h2"
              className="summary-title"
            >
              Total amount repayable: $
            </Header>
            <Header
              as="h2"
              className="summary-title"
            >
              Loan cost: $
              NaN
            </Header>
            <Header
              as="h2"
              className="summary-title"
            >
              Repayment date:
            </Header>
            <Button
              as="button"
              onClick={[Function]}
              primary={true}
            >
              Request loan
            </Button>
          </Fade>
          <Fade
            collapse={true}
            top={true}
            when={true}
          >
            <Header
              as="h2"
              className="summary-title"
            >
              Fill out your required loan details in the calculator
            </Header>
          </Fade>
        </GridColumn>
      </GridRow>
    </Grid>
  </Segment>
  <LoanHistory
    loanHistory={Array []}
    onLoanExtend={[Function]}
  />
  <Modal
    centered={true}
    closeOnDimmerClick={true}
    closeOnDocumentClick={false}
    dimmer={true}
    eventPool="Modal"
    onClose={[Function]}
  />
</div>
`;
